core/src/main/java/me/mykindos/betterpvp/core/client/punishments/PunishmentListener.java

```diff
@@ -118,4 +122,17 @@ public void onCanHurt(EntityCanHurtEntityEvent event) {
         }
     }
 
+    @UpdateEvent(delay = CHECKDELAY, isAsync = true)
+    public void onExpiration() {
+        long timeStart = System.currentTimeMillis();
+        clientManager.getOnline().forEach(client -> {
+            client.getPunishments().forEach(punishment -> {
+                if (punishment.isRevoked()) return;
+                if (punishment.hasExpired() && timeStart - punishment.getExpiryTime() <= CHECKDELAY) {
+                    punishment.getType().onExpire(client, punishment);
+                }
+            });
+        });
```

Comment:
Since the online client list isnt necessarily accurate (a player can go offline and still be considered online for some time afterwards due to caching)

You are probably better off just looping through online players instead

Reply:
That's right, changed